ngcc: webpack support

petebacondarwin opened the pull request:
This PR is a collection of fixes, refactorings and features for ngcc that should allow a webpack plugin to be created that triggers ngcc compilation.

Covers:

* FW-1118 ngcc - expose the programmatic API
* FW-1119 ngcc should be able to compile an individual entry-point (and its dependencies)
* FW-1120 ngcc should accept a fallback list of formats
* FW-1122 ngcc should attempt to use a given format to compute the DTS changes

gkalpak commented on scripts/build-packages-dist.sh:
```diff
@@ -1,6 +1,10 @@
+
+DIR="${BASH_SOURCE%/*}"
+if [[ ! -d "$DIR" ]]; then DIR="$PWD"; fi
+source "$DIR/package-builder.sh"
```
I have a similar commit in a WIP PR of mine, were I did (here and in [build-ivy-npm-packages.sh][1]):

```diff
-source ./scripts/package-builder.sh
+source $(dirname $0)/package-builder.sh
```

Isn't that enough? 

[1]: https://github.com/angular/angular/blob/aabc74c7fcf5a7bc70d1eb7474892339d5610116/scripts/build-ivy-npm-packages.sh

petebacondarwin replied:
I was basing this off StackOverflow expertise :-) It seems that `$dirname 0` can have issues...
https://stackoverflow.com/questions/192292/how-best-to-include-other-scripts/12694189#12694189

gkalpak replied:
Well, I think it will be fine for our usecase (and we use it elsewhere).
(We are not very consistent, though :grin: For example, we also use `$(realpath "$(dirname ${BASH_SOURCE[0]})")` 🤷‍♂️)

petebacondarwin commented on scripts/build-packages-dist.sh:
```diff
@@ -1,6 +1,10 @@
 #!/usr/bin/env bash
 
-source ./scripts/package-builder.sh
+echo ${BASH_SOURCE}
```
Obviously this line will disappear :-)

filipesilva commented on packages/compiler-cli/ngcc/test/packages/build_marker_spec.ts:
```diff
@@ -94,58 +94,49 @@ function restoreRealFileSystem() {
+      const pkg = JSON.parse(readFileSync('/node_modules/@angular/common/package.json', 'utf8'));
+      pkg.__modified_by_ngcc__ = {fesm2015: 'WRONG_VERSION'};
+      writeFileSync('/node_modules/@angular/common/package.json', JSON.stringify(pkg), 'utf8');
+      expect(() => checkMarker(createEntryPoint('/node_modules/@angular/common'), 'fesm2015'))
```
Is this error thrown up to the API user? How should it be handled?

petebacondarwin replied:
Good question - I think for your purposes it should be caught by the webpack plugin and surfaced to the user via whatever error reporting method webpack uses, right?

petebacondarwin replied:
I refactored the `mainNgcc()` function so that it does not swallow exceptions. The logging of these errors to the console is now done in the `ngcc-main.ts` command line script.

filipesilva replied:
SGTM

filipesilva commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -9,14 +9,14 @@
 export function mainNgcc(
     baseSourcePath: string, formats: EntryPointFormat[], baseTargetPath: string,
-    targetEntryPointPath?: string): number {
+    targetEntryPointPath?: string, propertiesToConsider?: EntryPointJsonProperty[]): number {
```
Are there tests for using `propertiesToConsider`?

petebacondarwin replied:
Not yet :-)

petebacondarwin replied:
Nor for `targetEntryPointPath` 😱

petebacondarwin replied:
There are integration tests for both now.

petebacondarwin commented on packages/compiler-cli/BUILD.bazel:
```diff
@@ -60,6 +60,6 @@ npm_package(
     ],
     deps = [
         ":compiler-cli",
-        "//packages/compiler-cli/src/ngcc",
```
That is an alternative but at the end of the day, ngcc is a separate tool from ngc, and so it doesn't really make sense for it to be a sub-folder of `compiler-cli/src`.

petebacondarwin commented:
Rebasing and adding integration tests.

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts:
```diff
@@ -1546,4 +1647,100 @@ describe('Esm5ReflectionHost', () => {
+         expect(internalClass2DtsDeclaration !.getSourceFile().fileName)
+             .toEqual('/typings/class2.d.ts');
+       });
+  });
```
I forgot to copy over the `getModuleWithProvidersFunctions()` tests. It looks like these are not working for ESM5.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1131,9 +1136,14 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
-        node && (ts.isFunctionDeclaration(node) || ts.isMethodDeclaration(node)) ? node : null;
+  protected parseForModuleWithProviders(
+      name: string, node: ts.Node|null, implementation: ts.Node|null = node,
+      container: ts.Declaration|null = null): ModuleWithProvidersFunction|null {
```
Oops I need to update the docs

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -57,51 +59,49 @@ export function mainNgcc({baseSourcePath, baseTargetPath = baseSourcePath, targe
 
+      if (checkMarker(entryPoint, property)) {
+        compiledFormats.add(formatPath);
+        continue;
```
Shouldn't we log [Skipping ${entryPoint.name} : ${property} (already compiled)][1] here?

[1]: https://github.com/angular/angular/pull/29092/commits/982638eaaf792b2a9a19f6c52db641afffc18733?file-filters%5B%5D=.ts#diff-2ec634b2f224017d93fb68397495e135R92

petebacondarwin replied:
Good spot. The logic in this section is really tricky.

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -57,51 +59,49 @@ export function mainNgcc({baseSourcePath, baseTargetPath = baseSourcePath, targe
+              `Skipping ${entryPoint.name} : ${format} (no valid entry point file for this format).`);
         }
+      } else {
+        console.warn(`Skipping ${entryPoint.name} : ${property} (already compiled).`);
```
I don't think we can ever end up here.

petebacondarwin replied:
I think we can...

This catches the scenario where two properties (e.g. `module` and `esm5`) happen to point to the same `formatPath` (e.g. `'./fesm/index.js'`).

In this case, there is not yet a processed marker for the second property, but the actual file has already been compiled. So we need to skip it (but still add the marker).

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -76,7 +81,7 @@ export function mainNgcc({baseSourcePath, targetEntryPointPath, propertiesToCons
       }
 
-      if (!compiledFormats.has(formatPath)) {
+      if (!compiledFormats.has(formatPath) && (compileAllFormats || compiledFormats.size === 0)) {
```
I find this condition a little confusing. I would rather `break` as soon as a format is compiled (and `compileAllFormats === false`).

gkalpak replied:
Oh, is this so that all properties that point to the same `formatPath` are marked as compiled?

petebacondarwin replied:
Right! I'll add a comment

gkalpak commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -31,50 +31,98 @@ describe('ngcc main()', () => {
-    });
-    expect(loadPackage('@angular/common/testing').__modified_by_ngcc__).toBeUndefined();
-    expect(loadPackage('@angular/common/http').__modified_by_ngcc__).toBeUndefined();
   });
```
It would be interesting to also check transitive dependencies (`a > b > c`).

gkalpak commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -31,50 +31,98 @@ describe('ngcc main()', () => {
+         });
+
+         // * the `main` property is UMD, which is not yet supported.
+         // * none fo the ES2015 formats are compiled as they are not on the `propertiesToConsider`
```
_fo_ --> _of_

gkalpak commented on integration/ngcc/test.sh:
```diff
@@ -10,11 +10,25 @@ ivy-ngcc --help
 # Did it add the appropriate build markers?
+
+  # - esm2015
+  grep '"__modified_by_ngcc__":[^}]*"esm2015":"' node_modules/@angular/common/package.json
```
> [^}]*

LoL - Backtracking FTW :smiley:

gkalpak commented on packages/compiler-cli/ngcc/main-ngcc.ts:
```diff
@@ -25,8 +25,7 @@ if (require.main === module) {
             array: true,
-            describe: 'An array of formats to compile.',
-            default: ['fesm2015', 'esm2015', 'fesm5', 'esm5']
+            describe: 'An array of package.json property formats to compile.'
```
This sounds confusing: "package.json property formats" 🤔
Maybe worth mentioning what happens if omitted (and which properties are supported - although I understand this entails the risk of becoming outdated without anyone noticing).

petebacondarwin replied:
I was trying to keep these description minimal. I'll add a little more, but we should probably have proper docs elsewhere - and also I believe that the majority of people should not be using these flags anyway :-)

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point.ts:
```diff
@@ -115,28 +94,72 @@ export function getEntryPointInfo(packagePath: string, entryPointPath: string):
+ * @param packageJsonPath the absolute path to the package.json file.
+ * @returns JSON from the package.json file if it is valid, `null` otherwise.
+ */
+export function loadEntryPointPackage(packageJsonPath: string): EntryPointPackageJson|null {
```
Why export this, if it is not being used elsewhere? Was the intention to use it in [build_marker][1]?

[1]: https://github.com/angular/angular/pull/29092/commits/949f779110a8f68ce7f952c2f386fefcc7dfb953?file-filters%5B%5D=.ts#diff-4840f884874632c2fbfdb64a26213396R58

petebacondarwin replied:
An abundance of transparency :-)

gkalpak commented on packages/compiler-cli/ngcc/src/packages/build_marker.ts:
```diff
@@ -7,35 +7,64 @@
-    }
+export function writeMarker(entryPoint: EntryPoint, format: EntryPointJsonProperty) {
+  const packageJsonPath = resolve(entryPoint.path, 'package.json');
+  const pkg = readPackage(packageJsonPath);
```
Why not use `entryPoint.packageJson`? Wouldn't it be faster?
Also, it would keep `entryPoint.packageJson` in sync with the actual `package.json` content.

petebacondarwin replied:
Excellent idea. When I originally wrote this commit the `entryPoint.packageJson` was not yet a thing.

gkalpak reviewed (changes requested):
💥 Amazing stuff 💥
(You almost make it look easy 😛)

Some commit message typos:
- a326a0236: _to use call ngcc_ --> _to use ngcc_
- 949f77911: _poth_ --> _path_
- 20ce8c9c5: _proces_ --> _process_
- 16f8a3d70: _the may be_ --> _there may be_

I think it would be good to have some `BREAKING CHANGE` notices in commits that change the public API, to help early adopters (in case they were using cli options).

petebacondarwin commented on packages/common/BUILD.bazel:
```diff
@@ -34,9 +34,9 @@ ng_package(
+        "//packages/compiler-cli/ngcc/test:__pkg__",
         "//packages/compiler-cli/test:__pkg__",
         "//packages/compiler-cli/test/diagnostics:__pkg__",
-        "//packages/compiler-cli/test/ngcc:__pkg__",
```
OK, so I want to keep the previous commit as a pure file move. So I will split this commit into the bits that update the files because of the move, and the bits that expose ngcc programmatically.

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -24,6 +24,11 @@ export interface NgccOptions {
+   * The path to the primary package to be compiled.
+   * All its dependencies will need to be compiled too.
+   */
+  targetEntryPointPath?: string;
```
I comes in a later refactoring :-)

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/entry_point_finder.ts:
```diff
@@ -18,9 +18,14 @@ export class EntryPointFinder {
+        targetEntryPointPath && path.resolve(sourceDirectory, targetEntryPointPath);
+    const targetEntryPoint = targetEntryPointPath ?
+        unsortedEntryPoints.find(entryPoint => entryPoint.path === targetEntryPointPath) :
+        undefined;
```
The first option doesn't work because `targetEntryPointPath` is a string and so we end up with the type of `targetEntryPoint` being `"" | EntryPoint | undefined`

The second option requires us to iterate the `find()` even if there is no chance of it finding anything.

I couldn't come up with a more succinct way of writing it than what is there. Any other ideas?

alxhub replied:
I prefer explicit to succinct in most cases. I like the logic as written.

petebacondarwin commented on aio/tools/examples/example-boilerplate.js:
```diff
@@ -12,80 +12,36 @@ const EXAMPLES_BASE_PATH = path.resolve(__dirname, '../../content/examples');
+    'src/assets/.gitkeep', 'src/browserslist', 'src/favicon.ico', 'src/karma.conf.js',
+    'src/polyfills.ts', 'src/test.ts', 'src/tsconfig.app.json', 'src/tsconfig.spec.json',
+    'src/tslint.json', 'e2e/src/app.po.ts', 'e2e/protractor.conf.js', 'e2e/tsconfig.e2e.json',
+    '.editorconfig', 'angular.json', 'package.json', 'tsconfig.json', 'tslint.json'
```
Yes auto-format :-( I'll see what I can do...

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -15,27 +18,30 @@ import {makeEntryPointBundle} from './packages/entry_point_bundle';
   /**
-   * Whether to compile all specified formats or to sstop compiling this entry-point at the first
+   * Whether to process all formats specified by (`propertiesToConsider`)  or to stop processing
+   * this entry-point at the first
```
No it is not ignored.

petebacondarwin replied:
The three options are independent of each other.

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/build_marker.ts:
```diff
@@ -14,18 +14,20 @@ import {EntryPoint, EntryPointJsonProperty} from './entry_point';
-  const compiledVersion = pkg && pkg.__modified_by_ngcc__ && pkg.__modified_by_ngcc__[format];
+export function hasBeenProcessed(packageJson: any, format: string) {
+  const compiledVersion =
+      packageJson && packageJson.__modified_by_ngcc__ && packageJson.__modified_by_ngcc__[format];
```
Hmm good point...

petebacondarwin replied:
Perhaps we should check that all the markers have the same version...

petebacondarwin replied:
Done

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -51,25 +57,28 @@ const SUPPORTED_FORMATS: EntryPointFormat[] = ['esm5', 'esm2015'];
 
-    const propertiesToCompile =
-        propertiesToConsider || Object.keys(entryPoint.packageJson) as EntryPointJsonProperty[];
+    const propertiesToCompile = propertiesToConsider || Object.keys(entryPoint.packageJson);
```
just use a fixed list of property names which is derived from the EntryPointJsonProperty type eh?

This is what I was asking you about recently - keeping a union type and an array of values in sync.

Despite that not being possible in TS, I think the benefit of not iterating over all the properties is worth it.

petebacondarwin replied:
I have refactored to only compute the value once per entry-point.

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -69,6 +69,16 @@ export function mainNgcc({basePath, targetEntryPointPath, propertiesToConsider,
+  if (absoluteTargetEntryPointPath && propertiesToConsider && entryPoints.every(entryPoint => {
+        return entryPoint.path !== absoluteTargetEntryPointPath;
+      })) {
+    // The requested entry-point did not get compiled so just marked it as processed
```
The comment is confusing sorry.

We would get here if the target entry-point was not an "Angular" entry-point, i.e. we did not find any `*.metadata.json` file at this entry-point. This means that it does not need to be compiled, and that is why our `findEntryPoints()` function does not include it.

petebacondarwin replied:
I need to add some tests for this situation, just to make sure we are not missing something.

petebacondarwin replied:
Currently, if `targetEntryPointPath` is provided, but it does not contain stuff for ngcc to do, it is marked...

but... if that entry-point has dependencies that do contain stuff for ngcc to process, they are not compiled.

I believe this is not a problem in the webpack scenario, since it will attempt to get ngcc to process every entry-point anyway, but in other scenarios it might be confusing...

petebacondarwin commented on packages/compiler-cli/ngcc/src/writing/in_place_file_writer.ts:
```diff
@@ -0,0 +1,36 @@
+ * a back-up of the original file with an extra `.bak` extension.
+ */
+export class InPlaceFileWriter implements FileWriter {
+  writeBundle(entryPoint: EntryPoint, _bundle: EntryPointBundle, transformedFiles: FileInfo[]) {
```
It should be `_entryPoint` too.

petebacondarwin commented on packages/compiler-cli/ngcc/test/BUILD.bazel:
```diff
@@ -34,6 +34,7 @@ jasmine_node_test(
         "//tools/testing:node_no_angular",
         "@npm//canonical-path",
         "@npm//convert-source-map",
+        "@npm//shelljs",
```
In `ngcc/BUILD.bazel` the dependency is used for typings at compile time.
Here the dependency is to ensure that the package is available at runtime (for testing).

gkalpak replied:
But if we use the package (in the source code), shouldn't it be listed as a dependency in `ngcc/BUILD.bazel`?
(You can tell I still don't fully understand bazel :grin:)

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -47,6 +48,10 @@ export interface NgccOptions {
+  /**
+   * Whether to create new entry-points bundles rather than overwriting the original files.
+   */
+  createNewEntryPointFormats?: boolean;
```
I was trying to avoid adding stuff to the CLI API if I can help it.
For the CLI I hope that people will mainly just run it over the whole node_modules folder.
We can always add stuff later :-)

petebacondarwin commented on packages/compiler-cli/ngcc/src/writing/new_entry_point_file_writer.ts:
```diff
@@ -0,0 +1,65 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {isDtsPath} from '@angular/compiler-cli/src/ngtsc/util/src/typescript';
```
Auto-import mistake :-)

petebacondarwin commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -84,22 +83,31 @@ interface Directory {
-function mockResolve(p: string): string|null {
-  if (existsSync(p)) {
-    const stat = statSync(p);
+function mockResolve(request: string, from: typeof Module): string|null {
```
When I was mucking about trying to fix this to make it work more like the real function, I added it.
But now I see that you are right it is just being passed around unnecessarily.

petebacondarwin commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -122,6 +122,40 @@ describe('ngcc main()', () => {
+
+      // Doesn't touch original files
+      expect(readFileSync(`${NODE_MODULES}/@angular/common/esm5/src/common_module.js`, 'utf8'))
+          .not.toContain(`import * as ɵngcc0 from '@angular/core';`);
```
Done.

petebacondarwin commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -122,6 +122,40 @@ describe('ngcc main()', () => {
+          .toContain(`import * as ɵngcc0 from '@angular/core';`);
+
+      // Copies over files that did not need compiling
+      expect(existsSync(`${NODE_MODULES}/@angular/common/__ivy_ngcc__/esm5/common.js`));
```
Done.

petebacondarwin commented on packages/compiler-cli/ngcc/test/writing/new_entry_point_file_writer_spec.ts:
```diff
@@ -0,0 +1,130 @@
+  mockFs({
+    '/node_modules/test-package': {
+      'secondary-entry-point': {
+        'package.json': '{"module": "index.js", "es2015": "es2015/index.js"}',
```
Ouch yes.

petebacondarwin replied:
I think I might need to change the algorithm so the new folder is at the root of the overall package rather than the entry-point...

petebacondarwin replied:
Done.

petebacondarwin commented:
Rebased and renamed the post-move commit to a fixup commit.
Over to you @gkalpak!

matsko commented:
Presubmit = http://test/OCL:239127076:BASE:239127199:1552972897129:7c1e3a24

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point.ts:
```diff
@@ -58,27 +47,27 @@ interface EntryPointPackageJson {
 
+export type EntryPointJsonProperty = keyof(PackageJsonFormatProperties);
+// We need to keep the elements of this const and the `EntryPointJsonProperty` type in sync.
+export const SUPPORTED_FORMAT_PROPERTIES: EntryPointJsonProperty[] =
```
I think `DEFAULT_FORMAT_PROPERTIES` would be more accurate. It's not like we don't support other properties.

petebacondarwin replied:
Actually we don't support other properties!! Take a look at `getEntryPointFormat()` if the property is not one of the `SUPPORTED_FORMAT_PROPERTIES` then we get `undefined`.

gkalpak replied:
Oh, good point :smiley:

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -0,0 +1,154 @@
+  if (absoluteTargetEntryPointPath && entryPoints.every(entryPoint => {
+        return entryPoint.path !== absoluteTargetEntryPointPath;
+      })) {
+    // If we get here, then the requested entry-point did not contain anything compile by
```
_compile by by_ --> _compiled by_ (?)

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -0,0 +1,154 @@
+
+  if (absoluteTargetEntryPointPath && entryPoints.every(entryPoint => {
+        return entryPoint.path !== absoluteTargetEntryPointPath;
+      })) {
```
I am still confused by this. How is it possible that `absoluteTargetEntryPointPath` should not be compiled but it has dependencies which should?
Could you give an example?

The reason I am worried is that we could be incorrectly marking `absoluteTargetEntryPointPath` as processed (e.g. if the user accidentally gave a path that is outside `basePath`).

petebacondarwin replied:
Consider a package that is just an aggregate of other packages:

**container-package/index.ts**
```ts
export * from 'package-1';
export * from 'package-2';
```
**container-package/package.json**
```
...
```
(Note no `*.metadata.json` file)

**package-1/index.ts**
```ts
import {Component} from '@angular/core';
@Component({...})
export class MyComponent {}
```

**package-1/package.json**
```
...
```

**package-1/index.metadata.json**
```
...
```

When compiling ngcc will note that `container-package` does not have a `metadata.json` file and so ignore it. If we set `targetEntryPointPath: 'container-package'` then we only try to compile that package and its dependencies. But since we ignored this package, all its dependencies get ignored too.  Therefore we do not compile anything.

gkalpak commented on packages/compiler-cli/ngcc/main-ngcc.ts:
```diff
@@ -0,0 +1,65 @@
+                'An array of names of properties in package.json to compile (e.g. `module` or `es2015`)\n' +
+                'Each of these properties should hold the path to a bundle-format.\n' +
+                'If provided, only the specified properties are considered for processing.\n' +
+                'If not provided, all the properties in the package.json are considered.'
```
No longer true :grin:

gkalpak reviewed (approved):
A couple of minor comments - nothing blocking.
Also, there is still a typo in 494679c8d: _poth_ --> _path_

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -0,0 +1,154 @@
+              `Skipping ${entryPoint.name} : ${format} (no valid entry point file for this format).`);
+        }
+      } else {
+        console.warn(`Skipping ${entryPoint.name} : ${property} (already compiled).`);
```
Good spot!

alan-agius4 commented:
Hi all, I just tested out using the transformed dts with a VE compilation on a basic library seems to be compiling as expected when using `createNewEntryPointFormats`.

Of course this will need further testing when having an application and more complex scenarios. But so far this looks promising.

alan-agius4 commented on packages/compiler-cli/ngcc/index.ts:
```diff
@@ -0,0 +1,17 @@
+ */
+
+import {hasBeenProcessed as _hasBeenProcessed} from './src/packages/build_marker';
+import {EntryPointJsonProperty, EntryPointPackageJson} from './src/packages/entry_point';
```
Shouldn't `NgccOptions` also be exported?

petebacondarwin replied:
Good spot!

petebacondarwin replied:
Fixed.

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -0,0 +1,154 @@
+const SUPPORTED_FORMATS: EntryPointFormat[] = ['esm5', 'esm2015'];
+
+/**
+ * This is the main entry-point into ngcc (aNgular Compatibility Compiler).
```
This is on purpose :-) But I realise I missed the G!
aNGular Compatibility Compiler => NGCC

alan-agius4 replied:
Ohhh

petebacondarwin replied:
Fixed

petebacondarwin committed:
feat(ivy): ngcc - support dts compilation via ES5 bundles

Previously we only compiled the typings files, in ngcc, if there was
an ES2015 formatted bundle avaiable. This turns out to be an artificial
constraint and we can also support typings compilation via ES5 formats
too.

This commit changes the ngcc compiler to attempt typings compilation
via ES5 if necessary. The order of the formats to consider is now:
FESM2015, FESM5, ESM2015, ESM5.

FW-1122

petebacondarwin committed:
test: ignore the generated `demo` folder in the `bazel` integration test

petebacondarwin committed:
style(ivy): ngcc - fix typo in comment

petebacondarwin committed:
build: allow build-packages-dist.sh to be run from anywhere

petebacondarwin committed:
build(ivy): ngcc - only test under `no-ivy-aot` mode

We want ngcc to test non-AOT builds of the Angular packages,
so we filter the tests in using the `no-ivy-aot` tag.

petebacondarwin committed:
refactor(ivy): move ngcc into a higher level folder

petebacondarwin committed:
refactor(ivy): ngcc - remove flat-format and use AbsoluteFsPath

Now that we are using package.json properties to indicate which
entry-point format to compile, it turns out that we don't really
need to distinguish between flat and non-flat formats, unless we
are compiling `@angular/core`.

petebacondarwin committed:
refactor(ivy): ngcc - simplify `Transformer.transform` API

By ensuring that EntryPointBundle contains everything that `Transformer.transform()`
needs to do its work, we can simplify its signature.

petebacondarwin committed:
feat(ivy): ngcc - support only compiling the first format property to match

By default ngcc will compile all the format properties specified. With this
change you can configure ngcc so that it will stop compiling an entry-point
after the first property that matches the `propertiesToConsider`.

petebacondarwin committed:
feat(ivy): ngcc - support targeting a start entry-point

You can now, programmatically, specify an entry-point where
the ngcc compilation will occur.
Only this entry-point and its dependencies will be compiled.

FW-1119

petebacondarwin committed:
feat(ivy): ngcc - compile only specified package.json format properties

You can now specify a list of properties in the package.json that
should be considered (in order) to find the path to the format to compile.

The build marker system has been updated to store the markers in
the package.json rather than an additional external file.
Also instead of tracking the underlying bundle format that was compiled,
it now tracks the package.json property.

BREAKING CHANGE:

The `proertiesToConsider` option replaces the previous `formats` option,
which specified the final bundle format, rather than the property in the
package.json.
If you were using this option to compile only specific bundle formats,
you must now modify your usage to pass in the properties in the package.json
that map to the format that you wish to compile.

In the CLI, the `--formats` is no longer available. Instead use the
`--properties` option.

FW-1120

petebacondarwin committed:
refactor(ivy): expose ngcc programmatically

The `mainNgcc()` function has been refactored to make it easier to call
ngcc from JavaScript, rather than via the command line.

For example, the `yargs` argument parsing and the exception
handling/logging have moved to the `main-ngcc.ts`
file so that it is only used for the command line version.

FW-1118

petebacondarwin committed:
fix(ivy): ngcc - empower `Esm5ReflectionHost` to analyze `ModuleWithProviders` functions

In ESM5 code, static methods appear as property assignments onto the constructor
function. For example:

```
var MyClass = (function() {
  function MyClass () {}
  MyClass.staticMethod = function() {};
  return MyClass;
})();
```

This commit teaches ngcc how to process these forms when searching
for `ModuleWithProviders` functions that need to be updated in the typings
files.

petebacondarwin committed:
refactor(ivy): ngcc - mark target entry-point as processed even if ngcc was a noop

If `targetEntryPointPath` is provided to `mainNgcc` then we will now mark all
the `propertiesToConsider` for that entry-point if we determine that
it does not contain code that was compiled by Angular (for instance it has
no `...metadata.json` file).

The commit also renames `__modified_by_ngcc__` to `__processed_by_ivy_ngcc__`, since
there may be entry-points that are marked despite ngcc not actually compiling anything.

petebacondarwin committed:
refactor(ivy): ngcc - extract file writing out into a class

This is in preparation of having different file writing strategies.

petebacondarwin committed:
feat(ivy): ngcc - support creating a new copy of the entry-point format

This commit adds a `NewEntryPointFileWriter` that will be used in
webpack integration. Instead of overwriting files in-place, this `FileWriter`
will make a copy of the TS program files and write the transformed files
there. It also updates the package.json with new properties that can be
used to access the new entry-point format.

FW-1121

petebacondarwin committed:
refactor(ivy): ngcc - clean up the public API and export `hasBeenProcessed` helper

Now the public API does not contain internal types, such as `AbsoluteFsPath` and
`EntryPointJsonProperty`. Instead we just accept strings and then guard them in
`mainNgcc` as appropriate.

A new public API function (`hasBeenProcessed`) has been exported to allow programmatic
checking of the build marker when the package.json contents are already known.

petebacondarwin committed:
fix(ivy): ngcc - fail build-marker check if any formats were compiled with different ngcc

Now we check the build-marker version for all the formats
rather than just the one we are going to compile.

This way we don't get into the situation where one format was
built with one version of ngcc and another format was built with
another version.

petebacondarwin committed:
fix(ivy): ngcc - teach Esm5ReflectionHost about aliased variables

Sometimes, in ESM5 code, aliases to exported variables are used internally
to refer to the exported value. This prevented some analysis from being
able to match up a reference to an export to the actual export itself.

For example in the following code:

```
var HttpClientXsrfModule = /** @class */ (function () {
  function HttpClientXsrfModule() {
  }
  HttpClientXsrfModule_1 = HttpClientXsrfModule;
  HttpClientXsrfModule.withOptions = function (options) {
      if (options === void 0) { options = {}; }
      return {
          ngModule: HttpClientXsrfModule_1,
          providers: [],
      };
  };
  var HttpClientXsrfModule_1;
  HttpClientXsrfModule = HttpClientXsrfModule_1 = tslib_1.__decorate([
      NgModule({
          providers: [],
      })
  ], HttpClientXsrfModule);
  return HttpClientXsrfModule;
}());
```

We were not able to tell that the `ngModule: HttpClientXsrfModule_1` property
assignment was actually meant to refer to the `function HttpClientXrsfModule()`
declaration.  This caused the `ModuleWithProviders` processing to fail.

This commit ensures that we can compile typings files using the ESM5
format, so we can now update the examples boilerplate tool so that it
does not need to compile the ESM2015 format at all.

petebacondarwin committed:
refactor(ivy): ngcc - use a fixed set of properties to compile if none provided

Previously we always considered all the properties in the package.json
if no `propertiesToConsidere` were provided.
But this results in computing a new set of properties for each entry-point
plus iterating through many of the package.json properties that are
not related to bundle-format paths.

petebacondarwin commented:
Caretaker note: ngcc is excluded from G3. But presubmit was run by @alxhub before the last rebase anyway.
